urlfd018a75f7_highlanderkev_StockExVirtuoso_tgz-pJ8-driver_Phase4J8/src/publishers/MessagePublisher.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * MessagePublisher.java
+ */
+package publishers;
+
+import client.User;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * This class implements the "singleton" Design pattern, there should only ever
+ * be one created instance.<br> This class sends out messages related to cancels
+ * or fills.
+ *
+ * @author Kevin Patrick Westropp
+ * @version StockExVirtuoso.Version.1.0
+ */
+public class MessagePublisher implements Publisher {
+
+    private static PublisherImpl myPublisher; // delegate - will refer to some implementation object 
+    private static MessagePublisher instance; // single instance of the MessagePublisher
+
+    private MessagePublisher() {
+    }
+
+    /**
+     * Public synchronized instance method to return instance of
+     * MessagePublisher, or if one has not been created, it creates one and
+     * returns it.
+     *
+     * @return MessagePublisher
+     */
+    public synchronized static MessagePublisher getInstance() {
+        if (instance == null) {
+            instance = new MessagePublisher();
+            myPublisher = new PublisherImpl();
+        }
+        return instance;
+    }
+
+    /**
+     * Public synchronized publish cancel messages to send out cancel messages
+     * to users.
+     *
+     * @param cm cancel message
+     */
+    public synchronized void publishCancel(CancelMessage cm) {
+        if (!myPublisher.getSubscribers().isEmpty()) {
+            HashMap<String, HashMap> subscribers = myPublisher.getSubscribers();
+            String userName = cm.getUser();
+            String product = cm.getProduct();
+            if (subscribers.get(product) != null) {
+                HashMap<String, User> users = subscribers.get(product);
+                if (users.get(userName) != null) {
+                    User u = (User) users.get(userName);
+                    u.acceptMessage(cm);
+                }
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 50:61, 71:82.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.